### ISSUE TITLE
Additional fields created by FlowDirection

### ISSUE THREAD

**Issue by kbarnhart** [landlab/landlab#714]
Right now FlowDirectors create two at-node fields `flow__receiver_node` and `flow__link_to_receiver_node`. Ideally, they would have an option to create other fields. 

A comprehensive list of fields might be:

@ link
`flow__direction` (-1, 0, 1) # indicates if flow direction is the same as topology direction
`flow__upstream_node`
`flow__downstream_node`
`flow__upstream_link_at_link`
`flow__downstream_link_at_link`

@ node:
`flow__link_to_receiver_node`
`flow__link_to_upstream_node`
`flow__receiver_node`
`flow__upstream_node`

**Comment by SiccarPoint** [comment]
`link_dirs_at_node` was designed so if you sum `[fluxes_on_links] * mg.link_dirs_at_node` for a given set of `links_at_node`, you get a net flux *INTO* the node. i.e., a negative number is assigned to any link pointing out of the node. So this means that where the node sits at the link TAIL, it gets a -1. I think this definition is consistent with what you have in parentheses, but not what you have outside them...???

Regardless, your proposed new definition 100% makes sense. It's just the sign on the fluxes that would be defined by the subsequent flow accumulation, right? I note this seems to be a fundamentally different thing from `link_dirs_at_node`: those are to do with link-node topology, but `flow__link_direction` is inherently a process definition...? (i.e., you could make a distributary network out of a catchment drainage by flipping `flow__link_direction`, but this would make zero difference to `link_dirs_at_node`)

**Comment by kbarnhart** [comment]
@SiccarPoint you are right. I reversed what is in the parentheses. Convention should be:

* 0 = No Flow
* -1 = Flow is going AGAINST the topology (flow going from head node to the tail node)
* 1 = Flow is going WITH the topology (flow going from the tail node to the head node)

Thanks for thinking about it carefully!

**Comment by nicgaspar** [comment]
@kbarnhart 
`flow__link_direction` sounds super helpful to me. I'd love if @jadams15 could chime in so let's give her a minute to respond too. She has thought about this SO much.

**Comment by jadams15** [comment]
I think there is a lot of benefit to having this - as Dan describes - as a process definition. As Nicole notes, I spent a lot of time thinking about ``link_dirs_at_node`` and often struggled to explain to new users how the link direction could be used to get at total discharge. It makes complete sense to me that for the network model you would move away from link direction and instead have a function for flow direction.

As long as the documentation is thorough and we keep link_dirs_at_node (my code relies on it!) I am 100% on board with these definitions and new methods!

**Comment by awickert** [comment]
Thanks @kbarnhart for the suggestions. This makes complete sense to me, and will make it easy for me to convert my own Network library (in progress, https://github.com/awickert/GRASS-fluvial-profiler/blob/master/v.stream.profiler/RiverNetwork.py) with LandLab. And thanks for hosting a very productive meeting at CSDMS as well!

**Comment by jadams15** [comment]
thanks @kbarnhart - I figured ``link_dirs_at_node`` would remain but I wanted to make doubly sure for obvious selfish reasons :)

Always happy to do a PR review! I meant to mention in my last comment but apparently omitted it - the reason I emphasize the documentation is that even with very thorough docs, I found it was really tricky to explain the link direction / topology thing to new users. Heck, I spent a lot of time thinking about this-  more than the average user - and I still had to draw out grids with 0, -1, 1 quite regularly (RIP my scrap paper pile). While in RasterModelGrid, most users don't need to really concern themselves with these things as they are handled under the hood (like in OverlandFlow), it seems especially important that users of the future NetworkModelGrid really digest the flow direction conventions. That said - your documentation is always awesome and clear so I am sure there will be no confusion once it's out there!

**Comment by kbarnhart** [comment]
@mcflugen: a technical detail...

`flow__link_direction` is number of links OR number of links + number of diagonals (e.g. number of d8s).

As best as I can tell we don't have a place on the grid to store things that are of this size. 

It appears to me that we have two options:
Option A: We permit a grid field that is `at_diagonals` and then data can be stored at diagonals.
Option B: These sorts of things are not stored as grid fields (they could be component attributes instead).

I don't really like B (but I'm at DIA at midnight... so I'm not sure I trust my judgement) because something like `flow__link_direction` is associated with each link and (when diagonals are considered) each diagonal. 

What do you think?

**Comment by SiccarPoint** [comment]
Good point Katy. I don't like B, but do like A. Perhaps with the proviso that we also have `at_d8`. Would be trivial to add syntactic sugar such that `at_link = at_d8[::2]`, `at_diagonal = at_d8[1:2]`. Am I recalling this ordering definition of d8 right? It's been a while.

Again though, we hit issues of back compatibility, and cross-grid compatibility. Any ideas Katy? You've played with these structures most recently, I think. My original way of sticking the diagonals on the end of the link lists was motivated purely by cross-grid compatibility: I wanted it so working on "links" in both grid types meant working on everything, and you didn't have to add extra "if then" code and grid type testing to work with a raster rather than a voronoi.

**Comment by kbarnhart** [comment]
I don't think we have severe back-compatibility issues or grid cross compatibility issues.
one part of LL that I'm less familiar with is @jadams15's  OverlandFlow component. Does it permit use of diagonals? Or store fluxes on links _and_ diagonals?

@mcflugen did a nice job cleaning up links, diagonals, and d8s, which makes this easier... having at_d8 would definitely make some things I'm working on easier. I'm nearly certain though that the ordering is 0:number_of_links-1, and then number_of_links:n_links + n_diagonals -1 (e.g. not interleaving links and diagonals as your sugar does). This also helpful b/c number of links and number of diagonals are not the same.

@mcflugen how would you feel about permitting at_diagonal and/or at_d8 grid fields for raster grids?
